src/core/row.cpp

```diff
@@ -119,5 +119,9 @@ blob row::move_as<blob>(std::size_t pos) const
 
     blob ret;
     type_conversion<blob>::move_from_base(baseVal, *indicators_.at(pos), ret);
+
+    // Re-initialize blob object so it can be used in further queries
+    baseVal.initialize(session);
```

Comment:
@vadz any idea how I could access a `session` object in this place? I need one as it is the only object that can create blob backend objects, which is required to get a fully functioning `blob` again.

An alternative I have considered is to write a `clone()` or `duplicate()` function for the `blob_backend` interface that can be used here (via some interface from the `blob` object) but that seems rather inconsistent with how things work for all other backend objects. Besides, that would essentially require duplicating the functionality of `session::make_blob_backend()` and code duplication is never a good thing...

Yet another alternative I was thinking of was to keep track of moved-from objects at the `row` level in order to re-initialize them only when the object is actually used again, but I don't think there currently exists a (nice) way of executing code before new data is bound to the row's holders (as they are simply referenced as `into` holders in the respective statement.

I'd appreciate thoughts from your side on this.

Reply:
This would require storing a back-pointer (or reference) to the session in each row object. It should probably be possible, but I wonder if we can avoid moving out of the blob here to avoid having to do this. Could you please remind me why was it necessary to add `move_as()` in the first place?

Reply:
Because blob objects are not copyable, which the standard `get` API does

Reply:
Yes, but couldn't we just provide access to their data instead?

Reply:
This would require copying data into some sort of buffer (plus getting the type conversion magic in SOCI's internals right), which might not be what the user wants. Just imagine the scenario in which I want to fetch a blob just to write it into a different table. That'd then require copying the data into my temp buffer just to yet again copy the data into a new `blob` object that I can then `use` in my insertion query.

Reply:
I don't know how common is this scenario and I think you could still do it using base API (and not rowset-based one)? So I'm really not sure if we want to complicate things just to support that...

Another possibility would be to implement copying for blobs, by using copy-on-write for them. But this also will add a lot of complexity. Maybe it's worth it, but I'm not sure.

Reply:
There's also the issue of having an easily understandable interface for users and given that SOCI has a dedicated `blob` object type, the expected thing (in my experience) is that one can fetch a `blob` object from a row (see also https://github.com/SOCI/soci/issues/922).
If you try that with the current code using `get<blob>` there will be a message telling you that you'll have to use `move_as<blob>`, which I suppose is still good enough in terms of users figuring out what to do with blobs.

So I kinda like the idea of having this functionality available. :eyes: 

A hack that I could think of would be adding a `session` object to the call signature of `move_as`:
```cpp
blob b = row.move_as<blob>(session);
```

I suppose that this wouldn't limit the usability for the user as they (presumably) have a `session` object at hand anyway when operating on `row`s :thinking:

Reply:
If you expect blobs to have value-like semantics, you have to implement support for copying them (which means using COW).

Passing `session` to `move_as()` is not great because it would be possible to pass a _different_ session here, which is something a well-designed API shouldn't allow.

Reply:
In my fork I use 
typedef std::shared_ptr<blob> blob_sptr; 

and 

        auto b = r.get<blob_sptr> ( 0 );
        CHECK ( b->get_len () == 10 );

Reply:
Pointers to blobs don't fix the issue, I think. You can still end up moving from the underlying blob elsewhere thereby breaking the `row` object again.

> If you expect blobs to have value-like semantics, you have to implement support for copying them (which means using COW).

Value-like as a move-only object. That's already fulfilled. And COW has the same underlying issue that we are trying to fix here, no? We need a way to create another blob backend, which can only be done via a `session` instance.

> Passing session to move_as() is not great because it would be possible to pass a different session here, which is something a well-designed API shouldn't allow.

True :thinking: 

How bad would adding a `session` member (as pointer) to `row` be? :thinking:

Reply:
> Pointers to blobs don't fix the issue, I think. You can still end up moving from the underlying blob elsewhere thereby breaking the `row` object again.
> 

This requires intentional user action. User could break a string, e.g.
string s;
sql.exchange(into(s));
string t = move(s);

Reply:
Yes but for `row` objects it is not at all obvious that the object the user receives is part of the internal state.

Reply:
I think the simplest would be to add `blob_impl::clone()`.

Reply:
Is it possible for the writable blob received from the select after insert ?

Reply:
> Is it possible for the writable blob received from the select after insert ?

Sorry, I can't quite follow - there seems to be a verb missing in your sentence (that's important for your meaning) :eyes:

Reply:
> > Is it possible for the writable blob received from the select after insert ?
> 
> Sorry, I can't quite follow - there seems to be a verb missing in your sentence (that's important for your meaning) 👀

Sorry, I mean blob_impl::clone. Is it possible to clone the blob? Who will own the blob locator ?

Reply:
Cloning is not a good idea, I think. We don't want to duplicate the stored data. But we can create a new, empty blob from an existing one.